reporter/src/main/kotlin/StatisticsCalculator.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.model.utils
+
+import com.here.ort.model.DependencyTreeStatistics
+import com.here.ort.model.IssueStatistics
+import com.here.ort.model.OrtResult
+import com.here.ort.model.Severity
+import com.here.ort.model.Statistics
+import com.here.ort.reporter.ResolutionProvider
+
+/**
+ * This class calculates [Statistics] for a given [OrtResult] and applicable
+ * [com.here.ort.model.config.ErrorResolution]s and applicable [com.here.ort.model.config.RuleViolationResolution]s.
+ */
+internal class StatisticsCalculator {
+    /**
+     * Return the [Statistics] for the given [ortResult].
+     */
+    fun getStatistics(ortResult: OrtResult, resolutionProvider: ResolutionProvider) = Statistics(
+        openIssues = getOpenIssuesIssues(ortResult, resolutionProvider),
+        openRuleViolations = getOpenRuleViolations(ortResult, resolutionProvider),
+        dependencyTree = DependencyTreeStatistics(
+            includedProjects = ortResult.getProjects().count { !ortResult.isExcluded(it.id) },
+            excludedProjects = ortResult.getProjects().count { ortResult.isExcluded(it.id) },
+            includedPackages = ortResult.getPackages().count { !ortResult.isExcluded(it.pkg.id) },
+            excludesPackages = ortResult.getPackages().count { ortResult.isExcluded(it.pkg.id) },
+            totalTreeDepth = getTreeDepth(ortResult),
+            includedTreeDepth = getTreeDepth(ortResult = ortResult, ignoreExcluded = true),
+            includedScopes = getIncludedScopes(ortResult).toSortedSet(),
+            excludedScopes = getExcludedScopes(ortResult).toSortedSet()
+        )
+    )
+
+    private fun getOpenRuleViolations(ortResult: OrtResult, resolutionProvider: ResolutionProvider): IssueStatistics {
+        val openPolicyViolations = ortResult
+            .getRuleViolations()
+            .filter { policyViolation -> resolutionProvider.getRuleViolationResolutionsFor(policyViolation).isEmpty() }
+
+        return IssueStatistics(
+            errors = openPolicyViolations.count { it.severity == Severity.ERROR },
+            warnings = openPolicyViolations.count { it.severity == Severity.WARNING },
+            hints = openPolicyViolations.count { it.severity == Severity.HINT }
```

Comment:
I believe we could avoid counting multiple times by using `groupBy` similar to here:

https://github.com/heremaps/oss-review-toolkit/blob/57adad65c32f4942a56221d2f60034f3d4438af6/reporter/src/main/kotlin/reporters/ReportTableModel.kt#L214-L220

This would also apply to the code below.

Reply:
That's funny, that's almost exactly what I had written before I optimized for readability.

Reply:
See above, I don't find the code we already have in `ReportTableModel` less readable at all.

Reply:
See above.